### PR TITLE
Retrieve current ManagedSpan to enable explicit release

### PR DESCRIPTION
**Pull request by objectiser**
Found an issue when trying to instrument Camel - the `SpanManager` returns a `ManagedSpan` when managing a span - but that span can only be released via the `ManagedSpan`.

If the instrumented application needs to `manage` and `release` in separate locations (i.e. separate callbacks on an exchange handler), then it is not possible to perform the release without stashing the ManagedSpan somewhere - which is the precise problem it is trying to solve.

This PR makes a slight change to the SpanManager API so that it returns a `ManagedSpan` when calling `currentSpan()` (method could be renamed, e.g. 'current()'?), allowing the code that finishes the span to then release it. The other approach that could be used would be to add a `release` method to the `SpanManager` to obtain the `refreshCurrent()` value and release it - which would be less of a change? Happy to discuss alternatives.

@sjoerdtalsma Comments appreciated.

### CODE REVIEW

**Comment by sjoerdtalsma** [comment]
I prefer adding a `current()` method that returns the current `ManagedSpan` instance (that is actually how my own [context-propagation](https://github.com/talsma-ict/context-propagation) lib works too 😉).

We can then deprecate the `currentSpan()` method and leave it in for one more revision to allow people time to adapt to this change.

**Comment by objectiser** [comment]
That is fine with me - I'll update the PR tomorrow.

**Comment by sjoerdtalsma** [comment]
@objectiser That would be great, thanks!

**Comment by sjoerdtalsma** [comment]
Maybe we should hold off on adding to the `SpanManager` interface until issue Should ManagedSpan become ManagedSpanContext (#5) settles?

**Comment by objectiser** [comment]
@sjoerdtalsma Thought I would just update based on your previous suggestion - just in case the decision is to keep using the Span instead of SpanContext.

**Comment by objectiser** [comment]
@sjoerdtalsma Given that there seems to be convergence in https://github.com/opentracing-contrib/java-spanmanager/issues/5 on propagating spans, is it ok to proceed with this PR?

**Comment by sjoerdtalsma** [comment]
@objectiser I think so; either way returning the `Managed` object (span or context) seems a good idea to me.

File: src/main/java/io/opentracing/contrib/spanmanager/SpanManager.java
Comment (sjoerdtalsma) [inline src/main/java/io/opentracing/contrib/spanmanager/SpanManager.java]: typo (extra ``<``)

**Review by sjoerdtalsma — APPROVED** [review]
Minor typo.
I think we need to think about what the 'empty' ManagedSpan means.
That discussion can be separate from this PR.

**Comment by objectiser** [comment]
@sjoerdtalsma Agree empty span needs to be considered separately based on the outcome of the discussions on https://github.com/opentracing/specification/issues/33.

btw could a 0.0.2 (or 0.1.0) release be done after this issue is merged?

**Comment by sjoerdtalsma** [comment]
@objectiser If it's up to me we could do a release soon.
The library is documented to be in-progress and the current behaviour continues to work (deprecated) for one more revision.

Since there haven't been any other participants in this PR yet, shall we give it 24h to see if others have any objections?

**Comment by bhs** [comment]
@sjoerdtalsma I like the idea of returning the `ManagedSpan` from `current()` for the reason you cited. I am less sure about returning a `noop` instance, though... is that needed for your use case somehow? If so, why? I can see it causing problems, as it becomes harder (IMO) than it used to be to figure out if there is or is not a managed span in the current thread.

**Comment by objectiser** [comment]
@bensigelman We have been discussing this issue in https://github.com/opentracing-contrib/java-spanmanager/issues/4 - agree null would be better if no managed span exists.

**Comment by objectiser** [comment]
@sjoerdtalsma As the `current` method is new, should we take this opportunity to return null if no managed span, and then deal with adding null childOf ref in the spec repo?

**Comment by sjoerdtalsma** [comment]
@objectiser That was my intention right after this merge :smile:

**Comment by objectiser** [comment]
@sjoerdtalsma Ok will update the PR to return null - do you want the commits squashed?

**Comment by sjoerdtalsma** [comment]
Github has the "Squash and merge" option, so whatever is easier for you.

**Comment by objectiser** [comment]
@sjoerdtalsma Good point :) - I'll leave to you. Hopefully this update should do it, but let me know if anything.

File: src/main/java/io/opentracing/contrib/spanmanager/SpanManager.java
Comment (sjoerdtalsma) [inline src/main/java/io/opentracing/contrib/spanmanager/SpanManager.java]: I think I misunderstood.
For semantics, I think we can return an EmptyManagedSpan / NoManagedSpan constant here, that contains a `null` Span.
That way for Java-8, we could add `Optional`-like methods to the interface providing a `getOrElse(Span)` or `ifPresent(Consumer<Span>)` default method.

I thought you would change the getSpan() to return `null`.

**Review by sjoerdtalsma — CHANGES_REQUESTED** [review]
I really don't like returning `null` ManagedSpan references. I have no objection to 'empty' managed spans that contain `null` Span references, but I would like `current()` to always return a ManagedSpan instance.
Let's document that it can be empty though!

File: src/main/java/io/opentracing/contrib/spanmanager/SpanManager.java
Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/SpanManager.java]: I am willing to do this separately from this PR too, just let me know.

**Comment by objectiser** [comment]
@sjoerdtalsma Sorry misunderstood - updated.

**Review by sjoerdtalsma — APPROVED** [review]

**Comment by objectiser** [comment]
@sjoerdtalsma Thanks! Would it be possible to do a release 0.0.2 or preferrably 0.1.0 today?

**Comment by sjoerdtalsma** [comment]
@objectiser 0.0.2 was released yesterday.
It is already available on maven-central: http://repo1.maven.org/maven2/io/opentracing/contrib/opentracing-spanmanager/0.0.2/

Maybe there's a mirror you use that was not yet synchronized?

**Comment by objectiser** [comment]
@sjoerdtalsma Ok thanks

**Pull request merged**
